controllers/installation_controller.go

```diff
@@ -312,102 +261,69 @@ func (r *InstallationReconciler) ReconcileHelmCharts(ctx context.Context, in *v1
 		return nil
 	}
 
-	if in.Status.State == v1beta1.InstallationStateFailed {
+	// skip if the installer has already completed, failed or if the k0s upgrade is still in progress
+	if in.Status.State == v1beta1.InstallationStateFailed ||
+		in.Status.State == v1beta1.InstallationStateInstalled ||
+		!in.Status.GetKubernetesInstalled() {
 		return nil
 	}
 
 	log := ctrl.LoggerFrom(ctx)
-	var clusterconfig k0sv1beta1.ClusterConfig
 	meta, err := release.MetadataFor(ctx, in.Spec.Config.Version, in.Spec.MetricsBaseURL)
 	if err != nil {
 		in.Status.SetState(v1beta1.InstallationStateHelmChartUpdateFailure, err.Error())
 		return nil
 	}
-	// skip if the new release has no addon configs
-	if meta.Configs == nil {
+
+	// skip if the new release has no addon configs - this should not happen in production
+	if meta.Configs == nil && in.Spec.Config.Extensions.Helm == nil {
 		log.Info("addons", "configcheck", "no addons")
 		if in.Status.State == v1beta1.InstallationStateKubernetesInstalled {
 			in.Status.SetState(v1beta1.InstallationStateInstalled, "Installed")
 		}
 		return nil
 	}
-	if in.Status.State == v1beta1.InstallationStateKubernetesInstalled || in.Status.State == v1beta1.InstallationStateHelmChartUpdateFailure {
-		var installedCharts k0shelm.ChartList
-		if err := r.List(ctx, &installedCharts); err != nil {
-			return fmt.Errorf("failed to list nodes: %w", err)
-		}
-		targetCharts := meta.Configs.Charts
-		chartErrors := []string{}
-		chartDrift := false
-		// grab the installed charts
-		for _, chart := range installedCharts.Items {
-			// extract any errors from installed charts
-			if chart.Status.Error != "" {
-				chartErrors = append(chartErrors, chart.Status.Error)
-			}
-			// check for version drift between installed charts and charts in the installer metadata
-			for _, targetChart := range targetCharts {
-				if targetChart.Name != chart.Status.ReleaseName {
-					continue
-				}
-				if targetChart.Version != chart.Spec.Version {
-					chartDrift = true
-				}
-			}
-		}
+
+	combinedConfigs := mergeHelmConfigs(meta, in)
+
+	// detect drift between the cluster config and the installer metadata
+	var installedCharts k0shelm.ChartList
+	if err := r.List(ctx, &installedCharts); err != nil {
+		return fmt.Errorf("failed to list installed charts: %w", err)
+	}
+	chartErrors, chartDrift := detectChartDrift(combinedConfigs, installedCharts)
+
+	// If all addons match their target version, mark installation as complete (or as failed, if there are errors)
+	if !chartDrift {
```

Comment:
this is an intentional logic change - now we will only mark the installation as 'charts have failed to install' if there aren't chart changes pending